### ISSUE TITLE
[FR] Suggest octal instead of hex escapes in `printf` format strings

### ISSUE THREAD

**Issue by Alhadis** [koalaman/shellcheck#2629]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:
```sh
#!/bin/sh
printf '\x41'
```

#### Here's what I wanted or expected to see:
~~~
In /tmp/example.sh line 2:
	printf '\x41'
	        ^--^ SC#### (warning): Use octal instead of hex escapes
	        ^--^ SC#### (warning): Replace 

Did you mean:
	printf '\101'
~~~

## Rationale
Character escape sequences of the form <code>\x<ins><var>HH</var></ins></code>, <code>\u<ins><var>HHHH</var></ins></code>, and <code>\U<ins><var>HHHHHHHH</var></ins></code> aren't portable in `printf` format strings:

~~~console
$ bash -c "printf '\\x41\n'"
A
$ dash -c "printf '\\x41\n'"
\x41
~~~

AFAIK, only octal escape sequences (<code>\\<ins><var>OOO</var></ins></code>) are sanctioned by POSIX in `printf` formats. Commands that recognise `printf`-like escapes (like `tr`) might be subject to similar portability hassles, although I haven't done too much research on this front.

**Comment by pietrmar** [comment]
I've run into this issue today and was definitely a bit disappointed that ShellCheck did not catch it, especially because it's already recommending replacing `echo -n -e` sequences with `printf`.

**Comment by e-kwsm** [comment]
FWIW dash 0.5.13.4 supports hex.

**Comment by Alhadis** [comment]
> FWIW dash 0.5.13.4 supports hex.

@e-kwsm The `printf` built-in of Dash v0.5.13 appears to support hexadecimal escapes, yes. But the `printf` of earlier <samp>dash(1)</samp> versions don't. Here's a test using Dash v0.5.12:

~~~console
λ ~: brew install --formula dash
==> Downloading https://ghcr.io/v2/homebrew/core/dash/manifests/0.5.12
==> Fetching dash
==> Downloading https://ghcr.io/v2/homebrew/core/dash/blobs/sha256:5f282ad1ebb1967545d5fd96625943ef81fa89be33487da251c7fd780bb22564
==> Pouring dash--0.5.12.monterey.bottle.tar.gz
🍺  /usr/local/Cellar/dash/0.5.12: 7 files, 227.9KB
λ ~: type dash
dash is /usr/local/bin/dash
λ ~: PS1='$ ' dash
$ printf '\x31\n'
\x31
~~~

In any case, while Dash pitches itself as a strictly-compliant POSIX shell, its man page readily admits it includes a few Berkeley extensions. I therefore don't consider it a reliable implementation of a <em>strictly</em> POSIX-compliant shell in the sense that, say, [<samp>posh(1)</samp>](https://salsa.debian.org/clint/posh) is.

(Also, ShellCheck has two different shell modes for `sh` and `dash`, so bringing the latter up here is kind of moot).

**Comment by e-kwsm** [comment]
posh does not have builtin `printf`:
https://salsa.debian.org/clint/posh/-/blob/c5b77889fee3a741cc61f2d15943471c7051a310/posh.1#L1477-1809
hence `/bin/printf` (GNU or BSD) is invoked.

But anyway warning should be issued when the specified shell is POSIX.

**Comment by Alhadis** [comment]
> posh does not have builtin `printf`:

I didn't say it did, only that I consider <samp>posh(1)</samp> a more stringent (and in the context of this discussion, more faithful) example of POSIX compliance than <samp>dash(1)</samp>.

In either case, I'm glad we're on the same page regarding my original request. 👍
